framework/src/task_manager/task_manager_core.c

```diff
@@ -316,52 +340,80 @@ static int taskmgr_start(int handle, int caller_pid)
 	return OK;
 }
 
+void taskmgr_update_stop_status(int signo, siginfo_t *data)
+{
+	int fd;
+	int ret;
+	int handle;
+	handle = taskmgr_get_handle_by_pid(data->si_value.sival_int);
+
+	/* Terminate based on task type */
+	fd = taskmgr_get_drvfd();
+	if (fd < 0) {
+		tmdbg("Invalid driver fd\n");
+	}
+
+	ret = ioctl(fd, TMIOC_TERMINATE, TM_PID(handle));
+	if (ret == ERROR) {
+		tmdbg("Fail to terminate the task\n");
+	}
+
+	/* task or pthread terminated well */
+	TM_STATUS(handle) = TM_APP_STATE_STOP;
+
+	TM_FREE(TM_STOP_CB_INFO(handle));
+	TM_STOP_CB_INFO(handle) = NULL;
+}
+
 static int taskmgr_stop(int handle, int caller_pid)
 {
 	int ret;
+	union sigval msg;
+	tm_termination_info_t info;
 
 	if (IS_INVALID_HANDLE(handle)) {
 		return TM_INVALID_PARAM;
 	}
 
 	ret = taskmgr_get_task_state(handle);
-	if (ret == TM_APP_STATE_UNREGISTERED || ret == TM_APP_STATE_STOP) {
+	if (ret == TM_APP_STATE_UNREGISTERED || ret == TM_APP_STATE_STOP || ret == TM_APP_STATE_CANCELLING) {
 		return -ret;
 	}
 
 	if (!taskmgr_is_permitted(handle, caller_pid)) {
 		return TM_NO_PERMISSION;
 	}
 
-
 	/* Call stop callback */
 	TM_STATUS(handle) = TM_APP_STATE_CANCELLING;
+
 	if (TM_STOP_CB_INFO(handle) != NULL) {
-		if (STOP_CBDATA(handle) != NULL) {
-			(*STOP_CBFUNC(handle))(STOP_CBDATA_MSG(handle));
-			TM_FREE(STOP_CBDATA_MSG(handle));
-			TM_FREE(STOP_CBDATA(handle));
-		} else {
-			(*STOP_CBFUNC(handle))(NULL);
-		}
-	}
+		info.cb = STOP_CBFUNC(handle);
+		info.cb_data = STOP_CBDATA(handle);
+		msg.sival_ptr = &info;
 
-	if (TM_TYPE(handle) == TM_BUILTIN_TASK || TM_TYPE(handle) == TM_TASK) {
-		ret = task_delete(TM_PID(handle));
-	}
+		ret = sigqueue(TM_PID(handle), SIGTM_TERMINATION, msg);
+		if (ret != OK) {
+			tmdbg("Fail to send signal, errno : %d\n", errno);
+			return TM_OPERATION_FAIL;
+		}
+	} else {
+		if (TM_TYPE(handle) == TM_BUILTIN_TASK || TM_TYPE(handle) == TM_TASK) {
+			ret = task_delete(TM_PID(handle));
+		}
 #ifndef CONFIG_DISABLE_PTHREAD
-	else {
-		(void)pthread_detach(TM_PID(handle));
-		ret = pthread_cancel(TM_PID(handle));
-	}
+		else {
+			(void)pthread_detach(TM_PID(handle));
+			ret = pthread_cancel(TM_PID(handle));
+		}
 #endif
-
-	if (ret != OK) {
-		tmdbg("Fail to delete the task\n");
-		return TM_OPERATION_FAIL;
+		if (ret != OK) {
+			tmdbg("Fail to delete the task\n");
+			return TM_OPERATION_FAIL;
+		}
+		/* task or pthread terminated well */
+		TM_STATUS(handle) = TM_APP_STATE_STOP;
```

Comment:
Is this right? Don't you use taskmgr_update_stop_status?

Reply:
This case is about not registering stop callback. In this case, status should be updated when stop the task.